java-util/src/main/java/io/druid/java/util/common/parsers/AbstractFlatTextFormatParser.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.java.util.common.parsers;
+
+import com.google.common.base.Function;
+import com.google.common.base.Optional;
+import com.google.common.base.Splitter;
+import com.google.common.base.Strings;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import io.druid.java.util.common.collect.Utils;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.IntFunction;
+
+public abstract class AbstractFlatTextFormatParser implements Parser<String, Object>
+{
+  public enum FlatTextFormat
+  {
+    CSV(","),
+    DELIMITED("\t");
+
+    private final String defaultDelimiter;
+
+    FlatTextFormat(String defaultDelimiter)
+    {
+      this.defaultDelimiter = defaultDelimiter;
+    }
+
+    public String getDefaultDelimiter()
+    {
+      return defaultDelimiter;
+    }
+  }
+
+  private final String listDelimiter;
+  private final Splitter listSplitter;
+  private final Function<String, Object> valueFunction;
+  private final boolean hasHeaderRow;
+  private final int maxSkipHeaderRows;
+  private final IntFunction<String> defaultColumnNameGenerator = ParserUtils.getDefaultColumnNameGenerator();
+
+  private List<String> fieldNames = null;
+  private boolean hasParsedHeader = false;
+  private int skippedHeaderRows;
+  private boolean supportSkipHeaderRows;
+
+  static Function<String, Object> getValueFunction(
+      final String listDelimiter,
+      final Splitter listSplitter
+  )
+  {
+    return (input) -> {
+      if (input.contains(listDelimiter)) {
+        return Lists.newArrayList(
+            Iterables.transform(
+                listSplitter.split(input),
+                ParserUtils.nullEmptyStringFunction
+            )
+        );
+      } else {
+        return ParserUtils.nullEmptyStringFunction.apply(input);
+      }
+    };
+  }
+
+  public AbstractFlatTextFormatParser(
+      final Optional<String> listDelimiter,
```

Comment:
According to #4275, maybe make it to accept simple String? I know subclasses also have Optional, but in a PR you can keep the extend of this problem, not increasing it.